### PR TITLE
CSP-1392: POST Permissions Endpoint

### PR DESCRIPTION
**Pull request by shaun-downey-education**
- Added a new endpoint to the **PermissionsController.cs** file to manage the adding a updating of provider permissions.

### CODE REVIEW

File: src/SFA.DAS.PR.Application/Permissions/Commands/PostPermissions/PostPermissionsCommandValidator.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace SFA.DAS.PR.Application.Permissions.Commands.PostPermissions;
+
+public class PostPermissionsCommandValidator : AbstractValidator<PostPermissionsCommand>
+{
+    public static readonly string UserRefValidationMessage = "A UserRef must be provided.";
+
+    public static readonly string UkprnValidationMessage = "A Ukprn must be provided.";
+
+    public static readonly string AccountLegalEntityIdValidationMessage = "A AccountLegalEntityId must be provided.";
```
Comment (MarkFCain) [inline src/SFA.DAS.PR.Application/Permissions/Commands/PostPermissions/PostPermissionsCommandValidator.cs:12]: 'An Account...'

File: src/SFA.DAS.PR.Application/SFA.DAS.PR.Application.csproj [hunk 2]
```diff
@@ -16,6 +16,7 @@
 	<ItemGroup>
 		<PackageReference Include="Azure.Identity" Version="1.11.2" />
 		<PackageReference Include="Microsoft.Identity.Client" Version="4.60.3" />
+		<PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Comment (MarkFCain) [inline src/SFA.DAS.PR.Application/SFA.DAS.PR.Application.csproj:19]: We're trying to not use newtonsoft, use System.Text.Json instead.  It's not difficult to swap, I\m happy to pair with you to get it done quickly

File: src/SFA.DAS.PR.Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandler.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        ).ToArray();
+
+        await permissionsWriteRepository.CreatePermissions(permissions, cancellationToken);
+        await CreatePermissionsAudit(command, command.Operations, PermissionAuditActions.PermissionCreatedAction, cancellationToken);
+        await providerRelationshipsDataContext.SaveChangesAsync(cancellationToken);
+
+        return new ValidatedResponse<PostPermissionsCommandResult>(new PostPermissionsCommandResult());
+    }
+
+    private async Task<ValidatedResponse<PostPermissionsCommandResult>> CheckExistingPermissions(AccountProviderLegalEntity accountProviderLegalEntity, PostPermissionsCommand command, CancellationToken cancellationToken)
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandler.cs:75]: Rename to `UpdatePermissions` so it captures all scenarios

File: src/SFA.DAS.PR.Data/EntityConfiguration/PermissionsAuditConfiguration.cs
[the remark this answers is not in the document]
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Data/EntityConfiguration/PermissionsAuditConfiguration.cs:1]: @DevenKShah I've tried removing it and it throws errors.

File: src/SFA.DAS.PR.Application/Permissions/Commands/PostPermissions/PostPermissionsCommandValidator.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        RuleFor(a => a.UserRef)
+            .NotEmpty()
+            .WithMessage(UserRefValidationMessage);
+
+        RuleFor(x => x.Ukprn)
+            .Cascade(CascadeMode.Stop)
+            .NotEmpty()
+            .WithMessage(UkprnValidationMessage);
+
+        RuleFor(x => x.Ukprn)
```
[the remark this answers is not in the document]
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Application/Permissions/Commands/PostPermissions/PostPermissionsCommandValidator.cs:25]: Previously tried this with Mark and it seems to ignore the initial error.

File: src/SFA.DAS.PR.Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandler.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+{
+    public async Task<ValidatedResponse<PostPermissionsCommandResult>> Handle(PostPermissionsCommand command, CancellationToken cancellationToken)
+    {
+        AccountProviderLegalEntity? accountProviderLegalEntity = await accountProviderLegalEntitiesReadRepository.GetAccountProviderLegalEntity(
+            command.Ukprn,
+            command.AccountLegalEntityId,
+            cancellationToken
+        );
+
+        if (accountProviderLegalEntity == null)
```
[the remark this answers is not in the document]
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandler.cs:29]: If the accountProviderLegalEntity is null then the logic specifies to add it

if (accountProviderLegalEntity == null)
{
    AccountLegalEntity? accountLegalEntity = await _accountLegalEntityReadRepository.GetAccountLegalEntity(
        command.AccountLegalEntityId,
        cancellationToken
    );

    if (accountLegalEntity == null)
    {
        return new ValidatedResponse<PostPermissionsCommandResult>(new PostPermissionsCommandResult());
    }

    return await CreateAccountProviderAndAddPermissions(command, accountLegalEntity.AccountId, cancellationToken);
}

File: src/SFA.DAS.PR.Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandler.cs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+            return await CreateAccountProviderAndAddPermissions(command, accountLegalEntity.AccountId, cancellationToken);
+        }
+
+        return await CheckExistingPermissions(accountProviderLegalEntity, command, cancellationToken);
+    }
+
+    private async Task<ValidatedResponse<PostPermissionsCommandResult>> CreateAccountProviderAndAddPermissions(PostPermissionsCommand command, long accountId, CancellationToken cancellationToken)
+    {
+        AccountProvider accountProvider = await accountProviderWriteRepository.CreateAccountProvider(
```
[the remark this answers is not in the document]
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandler.cs:49]: We spoke on this and agreed that separating concerns would be the preference

File: src/SFA.DAS.PR.Api.UnitTests/Controllers/Permissions/PermissionsControllerGetTests.cs
Comment (DevenKShah) [inline src/SFA.DAS.PR.Api.UnitTests/Controllers/Permissions/PermissionsControllerGetTests.cs:1]: I see you have changed namespace, but now the folder structure does not match, can you rename the folder as well please

File: src/SFA.DAS.PR.Api.UnitTests/Controllers/Permissions/PermissionsControllerTests.cs
Comment (DevenKShah) [inline src/SFA.DAS.PR.Api.UnitTests/Controllers/Permissions/PermissionsControllerTests.cs:1]: Renames this file to match the convention used in the other two, ie. PermissionsControllerHasRelationshipWithPermissionsTest

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**